gst/nnstreamer/nnstreamer_subplugin.c

```diff
@@ -0,0 +1,206 @@
+_strequal (gconstpointer _a, gconstpointer _b)
+{
+  const subpluginData *a = _a, *b = _b;
+  return (0 == g_strcmp0 (a->name, b->name));
+}
+
+/** @brief Public function defined in the header */
+const void *
+get_subplugin (subpluginType type, const char *name)
+{
+  GHashTable *table;
+  subpluginData *data;
+  subpluginData key;
+
+  if (subplugins[type] == NULL)
+    subplugins[type] = g_hash_table_new (_sphash, _strequal);
+
+  table = subplugins[type];
+
+  key.name = g_strdup (name);
+
+  data = g_hash_table_lookup (table, &key);
+
+
+  if (data == NULL) {
+    /* Search and register if found with the conf */
+    const gchar *fullpath = nnsconf_get_fullpath (name, type);
+    void *handle;
+    char *dlsym_error;
+    nnstreamer_subplugin_data *nsdata;
+
+    if (fullpath == NULL)
+      goto error;               /* No Such Thing !!! */
+
+    handle = dlopen (fullpath, RTLD_NOW);
+    if (NULL == handle) {
+      GST_ERROR ("Cannot dlopen %s (%s).", name, fullpath);
+      goto error;
+    }
+
+    nsdata = (nnstreamer_subplugin_data *)
+        dlsym (handle, "nnstreamer_subplugin");
+    dlsym_error = dlerror ();
+
+    if (NULL == nsdata) {
+      GST_ERROR ("nnstreamer_subplugin does not exists in %s (%s)", name,
+          fullpath);
+      dlclose (handle);
+      goto error;
+    }
+    if (dlsym_error) {
+      GST_ERROR ("Loading nnstreamer_subplugin in %s (%s) incurs: %s", name,
+          fullpath, dlsym_error);
+      dlclose (handle);
+      goto error;
+    }
+
+    if (nsdata->checker != NNS_SUBPLUGIN_CHECKER) {
+      GST_ERROR ("nnstreamer_subplugin of %s (%s) is broken (first bytes)",
+          name, fullpath);
+      dlclose (handle);
+      goto error;
+    }
+
+    if (nsdata->type != type) {
+      GST_ERROR ("nnstreamer_subplugin of %s (%s) is broken (type mismatch)",
+          name, fullpath);
+      dlclose (handle);
+      goto error;
+    }
+
+    if (g_strcmp0 (nsdata->name, name)) {
+      GST_ERROR ("nnstreamer_subplugin of %s (%s) is broken (name mismatch)",
+          name, fullpath);
+      dlclose (handle);
+      goto error;
+    }
+
+    g_assert (TRUE == register_subplugin (type, name, data));
+    g_assert ((data = g_hash_table_lookup (table, &key)) != NULL);
+    data->handle = handle;
+  }
+
+  g_free (key.name);
+  return data->data;
+error:
+  g_free (key.name);
+  return NULL;
+}
+
+/** @brief Public function defined in the header */
+gboolean
+register_subplugin (subpluginType type, const char *name, const void *data)
+{
+  /** @todo data out of scope at add */
+  subpluginData *spdata = g_new (subpluginData, 1);
+  gboolean ret = TRUE;
+
+  spdata->name = g_strdup (name);
+  spdata->data = data;
+  spdata->handle = NULL;
+
+  if (subplugins[type] == NULL)
+    subplugins[type] = g_hash_table_new (_sphash, _strequal);
+
+  G_LOCK (splock);
+  if (g_hash_table_contains (subplugins[type], spdata))
+    ret = FALSE;
+
+  if (ret == TRUE)
+    g_hash_table_add (subplugins[type], spdata);
+
```

Comment:
I'm wondering g_hash_table_add() is always successful? I think if it does not, then we may need free spdata.

Reply:
It's changed to g_hash_table_insert(), which does not "FAIL-RETURN".
It returns FALSE if the key already exists; however, it simply overwrites if so.
Thus, we do not need to worry about spdata. We just need to worry about the overwritten old data.

Reply:
And the old data is now handled by _spdata_destroy